🐛  Start the Cache if the Manager has already started

jsanda opened the pull request:
<!-- What does this do, and why do we need it? -->
Fixes #1673. 

This is needed to better support multi-cluster controllers where you want to dynamically add Clusters after the Manager has already started and have their caches started and synced.

jsanda commented:
@alvaroaleman if my changes look reasonable I can add a new test in manager_test.go.

alvaroaleman commented on pkg/manager/internal.go:
```diff
@@ -211,6 +211,12 @@ func (cm *controllerManager) Add(r Runnable) error {
 		cm.caches = append(cm.caches, hasCache)
+		if cm.started {
+			cm.startRunnable(hasCache)
+			if !hasCache.GetCache().WaitForCacheSync(cm.internalCtx) {
```
This will never timeout, right? Does that match the standard path, where we start the caches before the `cm` is started?

/ok-to-test

jsanda replied:
It could definitely time out. The remote cluster might not be accessible. Should the cache be started and synced with a different context? If the sync times out, I would then cancel the context. I suppose it also makes more sense to append the cache after the sync completes successfully.

I believe this matches the standard path.  I can't simply call `waitForCache` because it returns immediately if `cm.started` is true. It first iterates through the caches and calls `cm.startRunnable` for each one. Then it iterates through the caches again and calls `cache.GetCache().WaitForCacheSync(ctx)` where `ctx` is `cm.internalCtx`.

jsanda replied:
After looking through the code some more,  and the `WaitForCacheSync` function in `shared_informer.go` in client-go in particular, I am less certain that it will time out.

jsanda commented:
@alvaroaleman I added a test. I added a fake cluster impl as a hook to verify the cluster is started. It feels a bit clunky. Open to suggestions if you have a better/simpler approach.

alvaroaleman commented:
thanks!
